app/assets/javascripts/pretty-text/engines/discourse-markdown/hashtag-autocomplete.js

```diff
@@ -0,0 +1,132 @@
+// NOTE: For future maintainers, the hashtag lookup here does not take
+// into account mixed contexts -- for instance, a chat quote inside a post
+// or a post quote inside a chat message, so hashtagTypesInPriorityOrder may
+// not provide an accurate lookup for hashtags without a ::type suffix in those
+// cases if there are conflcting types of resources with the same slug.
+
+function addHashtag(buffer, matches, state) {
+  const options = state.md.options.discourse;
+  const slug = matches[1];
+  const hashtagLookup = options.hashtagLookup;
+
+  // NOTE: The lookup function is only run when cooking
+  // server-side, and will only return a single result based on the
+  // slug lookup.
+  const result =
+    hashtagLookup &&
+    hashtagLookup(
+      slug,
+      options.currentUser,
+      options.hashtagTypesInPriorityOrder
+    );
+
+  // NOTE: When changing the HTML structure here, you must also change
+  // it in the placeholder HTML code inside lib/hashtag-autocomplete, and vice-versa.
+  let token;
+  if (result) {
+    token = new state.Token("link_open", "a", 1);
```

Comment:
Any chance we can simplify all this code building the HTML as a string like we do [here](https://github.com/discourse/discourse/blob/main/app/assets/javascripts/pretty-text/engines/discourse-markdown/image-controls.js#L112)?

Reply:
I would prefer to leave it as the token builder, most of our markdown rules are done that way. I don't see much of a benefit in changing it now, though I will keep this in mind for future rules